assets/js/state/sagas/hosts.js

```diff
@@ -1,7 +1,54 @@
-import { put, takeEvery } from 'redux-saga/effects';
-import { HOST_DEREGISTERED, removeHost } from '@state/hosts';
+import {
+  all,
+  delay,
+  put,
+  race,
+  call,
+  take,
+  takeEvery,
+} from 'redux-saga/effects';
+import {
+  DEREGISTER_HOST,
+  CANCEL_DEREGISTER_HOST,
+  HOST_DEREGISTERED,
+  removeHost,
+  setHostDeregisterable,
+} from '@state/hosts';
 import { notify } from '@state/actions/notifications';
 
+export function* markDeregisterableHosts(hosts) {
+  yield all(
+    hosts
+      .filter(({ heartbeat }) => heartbeat !== 'passing')
+      .map((host) =>
+        put({
+          type: DEREGISTER_HOST,
+          payload: host,
+        })
+      )
+  );
+}
+
+function* deregisterHostDetail(payload) {
+  // eslint-disable-next-line no-undef
+  yield delay(config.deregistrationDebounce ?? 0);
+  yield put(setHostDeregisterable(payload));
+}
+
+function* deregisterHost({ payload }) {
+  yield race({
+    response: call(deregisterHostDetail, payload),
+    cancel: take(
+      ({ type, payload: cancelPayload }) =>
+        type === CANCEL_DEREGISTER_HOST && cancelPayload.id === payload.id
```

Comment:
If we extract this to an actual function, we could test it, and skip the `race` testing as it is difficult. But testing this `matching` function looks worthy